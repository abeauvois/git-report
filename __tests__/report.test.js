const { from, of, combineLatest } = require("rxjs");
const { toArray, skip, map, take, mergeMap, flatMap, tap } = require("rxjs/operators");

const { TestScheduler } = require("rxjs/testing");

const { fromCSVFile } = require("../src/fromCSVFile");
const { TasksManager } = require("../src/TasksManager");

const {
  getTask$,
  aggregateDiffs,
  mergeTasksWithCalendar,
  groupAndMergeByYear,
  groupAndMergeByMonth,
  groupAndMergeByDay,
  groupMessagesByDay,
  buildReport,
  sendReport,
  groupAndMergeByWeek,
  getBounds,
} = require("../src/report");

describe("report", () => {
  let rxTest;
  const filename = "./__tests__/test_file.csv";
  const filenameR2 = "./__tests__/test_file_report_R2_b.csv";
  const filenameR3a = "./__tests__/test_file_report_R3_a.csv";
  const filenameR6b = "./__tests__/test_file_report_R6_b.csv";

  beforeEach(() => {
    rxTest = new TestScheduler((actual, expected) => {
      expect(actual).toEqual(expected);
    });
  });

  test("R1.a: fromCSVFile", (done) => {
    fromCSVFile(filename)
      .pipe(take(2))
      .subscribe(
        (received) => {
          expect(received).toMatchSnapshot();
        },
        (e) => console.log("Errroor", e),
        () => done()
      );
  });

  test("R1.b: getTask$", (done) => {
    getTask$(filename).subscribe(
      (received) => {
        expect(received).toMatchSnapshot();
      },
      (e) => console.log("Errroor", e),
      () => done()
    );
  });

  test("R2.a: mergeTasksWithCalendar should start on firstTaskDate", (done) => {
    const firstTaskDate = "2020/11/12";
    const lastTaskDate = "2020/11/13";

    const task$ = getTask$(filenameR2).pipe(
      flatMap((tasks) => {
        const tm = TasksManager(tasks);
        return of(tm);
      })
    );

    mergeTasksWithCalendar(task$).subscribe(
      (received) => {
        expect(received[0].date).toEqual(firstTaskDate);
        expect(received[0].message).toEqual("off");

        expect(received.length).toEqual(4);
        expect(received[received.length - 1].date).toEqual(lastTaskDate);
        expect(received[received.length - 1].message).toEqual("m1 13 nov");
      },
      (e) => console.log("Errroor", e),
      () => done()
    );
  });

  test("R2.b: mergeTasksWithCalendar should start 1 day before lastTaskDate", (done) => {
    const lastTaskDate = "2020/11/13";

    const takeLast = 1;

    const task$ = getTask$(filenameR2).pipe(
      flatMap((tasks) => {
        const tm = TasksManager(tasks);
        return of(tm);
      })
    );

    mergeTasksWithCalendar(task$, takeLast).subscribe(
      (received) => {
        expect(received[0].date).toEqual(lastTaskDate);
        expect(received[0].message).toEqual("off");

        expect(received.length).toEqual(2);
        expect(received[received.length - 1].date).toEqual(lastTaskDate);
        expect(received[received.length - 1].message).toEqual("m1 13 nov");
      },
      (e) => console.log("Errroor", e),
      () => done()
    );
  });

  test("R3.a: groupByYear", (done) => {
    const task$ = getTask$(filenameR2).pipe(
      flatMap((tasks) => {
        const tm = TasksManager(tasks);
        return of(tm);
      })
    );

    mergeTasksWithCalendar(task$)
      .pipe(
        flatMap((tasks) => {
          return from(tasks);
        }),
        groupAndMergeByYear
      )
      .subscribe(
        (received) => {
          expect(received).toMatchSnapshot();
        },
        (e) => console.log("Errroor", e),
        () => done()
      );
  });

  test("R3.b: aggregateDiffs ", () => {
    rxTest.run((helpers) => {
      const { cold, expectObservable } = helpers;

      const days = {
        a: {
          ampm: "am",
          day: 1,
          dayWeek: 5,
          hour: 1,
          message: "meeting",
          milliseconds: 1609459200000,
          month: 0,
          year: 2021,
        },
        b: {
          ampm: "pm",
          day: 4,
          dayWeek: 3,
          hour: 13,
          message: "meeting",
          milliseconds: 1635943239000,
          month: 10,
          year: 2021,
        },
      };
      const tasks = {
        i: {
          ampm: "pm",
          day: 3,
          dayWeek: 3,
          hour: 13,
          message: "m1 03 nov 21",
          milliseconds: 1635943239000,
          month: 10,
          year: 2021,
        },
        j: {
          ampm: "pm",
          day: 4,
          dayWeek: 3,
          hour: 13,
          message: "m1 04 nov 21",
          milliseconds: 1635943239000,
          month: 10,
          year: 2021,
        },
      };
      const day$ = cold("-a--b|", days);
      const task$ = cold("-i--j|", tasks);

      const expectedMarble = "-n--m|";
      const expectedValues = {
        n: [tasks.i, days.a],
        m: [tasks.j],
      };

      const received = aggregateDiffs(task$, day$);
      expectObservable(received).toBe(expectedMarble, expectedValues);
    });
  });

  test("R4: groupMessagesByDay", (done) => {
    const task$ = getTask$(filenameR2).pipe(
      flatMap((tasks) => {
        const tm = TasksManager(tasks);
        return of(tm);
      })
    );

    mergeTasksWithCalendar(task$)
      .pipe(
        flatMap((tasks) => {
          return from(tasks);
        }),
        groupAndMergeByDay,
        mergeMap(groupMessagesByDay),
        toArray()
      )
      .subscribe(
        (received) => {
          expect(received).toMatchSnapshot();
        },
        (e) => console.log("Errroor", e),
        () => done()
      );
  });

  test("R5: groupByYear", (done) => {
    const task$ = getTask$(filenameR3a).pipe(
      flatMap((tasks) => {
        const tm = TasksManager(tasks);
        return of(tm);
      })
    );

    const source$ = mergeTasksWithCalendar(task$).pipe(
      flatMap((tasks) => {
        return from(tasks);
      }),
      groupAndMergeByDay,
      mergeMap(groupMessagesByDay),
      groupAndMergeByYear,
      flatMap(([year, tasks]) => combineLatest([of(year), from(tasks).pipe(groupAndMergeByMonth)])),
      flatMap(([year, [month, tasks]]) =>
        combineLatest([of(year), combineLatest([of(month), from(tasks).pipe(groupAndMergeByWeek)])])
      )
    );

    source$.pipe(tap((t) => console.log("t:", t))).subscribe(
      (received) => {
        expect(received).toMatchSnapshot();
      },
      (e) => console.log("Errroor", e),
      () => done()
    );
  });

  test("R6.a: buildReport", (done) => {
    buildReport({ filename: filenameR6b }).then((received) => {
      expect(received).toMatchSnapshot();
      done();
    });
  });

  test("R6.b: buildReport last 2 days", (done) => {
    buildReport({ filename: filenameR6b, lastDays: 2 }).then((received) => {
      expect(received).toMatchSnapshot();
      done();
    });
  });

  test.skip("R8: sendReport to console", (done) => {
    sendReport({ filename: filenameR3a, channel: "console" }).then((received) => {
      expect(received).toMatchSnapshot();
      done();
    });
  });
});
