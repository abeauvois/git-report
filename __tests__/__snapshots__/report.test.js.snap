// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`report R1.a: fromCSVFile 1`] = `
Object {
  "contributor": " contributor",
  "date": " date",
  "message": " message",
  "sha": "sha",
}
`;

exports[`report R1.a: fromCSVFile 2`] = `
Object {
  "contributor": " abeauvois",
  "date": " Thu Nov 12 16:06:45 2020",
  "message": " \\"m2 12 nov PM\\"",
  "sha": "3",
}
`;

exports[`report R1.b: getTask$ 1`] = `
Array [
  Object {
    "ampm": "pm",
    "date": "2020/09/12",
    "day": 12,
    "dayWeek": 6,
    "hour": 16,
    "message": "m1 sep",
    "milliseconds": 1599919401000,
    "month": 8,
    "week": 37,
    "year": 2020,
  },
  Object {
    "ampm": "pm",
    "date": "2020/10/12",
    "day": 12,
    "dayWeek": 1,
    "hour": 16,
    "message": "m2 oct",
    "milliseconds": 1602511401000,
    "month": 9,
    "week": 42,
    "year": 2020,
  },
  Object {
    "ampm": "am",
    "date": "2020/11/12",
    "day": 12,
    "dayWeek": 4,
    "hour": 10,
    "message": "m3 12 nov AM",
    "milliseconds": 1605171801000,
    "month": 10,
    "week": 46,
    "year": 2020,
  },
  Object {
    "ampm": "am",
    "date": "2020/11/12",
    "day": 12,
    "dayWeek": 4,
    "hour": 11,
    "message": "m3 12 nov AM",
    "milliseconds": 1605175401000,
    "month": 10,
    "week": 46,
    "year": 2020,
  },
  Object {
    "ampm": "pm",
    "date": "2020/11/12",
    "day": 12,
    "dayWeek": 4,
    "hour": 16,
    "message": "m2 12 nov PM",
    "milliseconds": 1605193605000,
    "month": 10,
    "week": 46,
    "year": 2020,
  },
  Object {
    "ampm": "am",
    "date": "2020/11/13",
    "day": 13,
    "dayWeek": 5,
    "hour": 10,
    "message": "m1 13 nov",
    "milliseconds": 1605260439000,
    "month": 10,
    "week": 46,
    "year": 2020,
  },
  Object {
    "ampm": "am",
    "date": "2021/01/01",
    "day": 1,
    "dayWeek": 5,
    "hour": 10,
    "message": "m1 01 janv 21",
    "milliseconds": 1609494039000,
    "month": 0,
    "week": 0,
    "year": 2021,
  },
  Object {
    "ampm": "pm",
    "date": "2021/02/01",
    "day": 1,
    "dayWeek": 1,
    "hour": 13,
    "message": "m1 01 fev 21",
    "milliseconds": 1612183239000,
    "month": 1,
    "week": 5,
    "year": 2021,
  },
  Object {
    "ampm": "pm",
    "date": "2021/11/03",
    "day": 3,
    "dayWeek": 3,
    "hour": 13,
    "message": "m1 03 nov 21",
    "milliseconds": 1635943239000,
    "month": 10,
    "week": 44,
    "year": 2021,
  },
]
`;

exports[`report R3.a: groupByYear 1`] = `
Array [
  2020,
  Array [
    Object {
      "ampm": "am",
      "date": "2020/11/12",
      "day": 12,
      "dayWeek": 4,
      "hour": 0,
      "message": "off",
      "milliseconds": 1605135600000,
      "month": 10,
      "week": 46,
      "year": 2020,
    },
    Object {
      "ampm": "am",
      "date": "2020/11/12",
      "day": 12,
      "dayWeek": 4,
      "hour": 10,
      "message": "m1 12 nov",
      "milliseconds": 1605174039000,
      "month": 10,
      "week": 46,
      "year": 2020,
    },
    Object {
      "ampm": "am",
      "date": "2020/11/13",
      "day": 13,
      "dayWeek": 5,
      "hour": 0,
      "message": "off",
      "milliseconds": 1605222000000,
      "month": 10,
      "week": 46,
      "year": 2020,
    },
    Object {
      "ampm": "am",
      "date": "2020/11/13",
      "day": 13,
      "dayWeek": 5,
      "hour": 10,
      "message": "m1 13 nov",
      "milliseconds": 1605260439000,
      "month": 10,
      "week": 46,
      "year": 2020,
    },
  ],
]
`;

exports[`report R4: groupMessagesByDay 1`] = `
Array [
  Object {
    "date": "2020/11/12",
    "day": 12,
    "dayWeek": 4,
    "messagesAm": "m1 12 nov",
    "messagesPm": "off",
    "month": 10,
    "week": 46,
    "year": 2020,
  },
  Object {
    "date": "2020/11/13",
    "day": 13,
    "dayWeek": 5,
    "messagesAm": "m1 13 nov",
    "messagesPm": "off",
    "month": 10,
    "week": 46,
    "year": 2020,
  },
]
`;

exports[`report R5: groupByYear 1`] = `
Array [
  2020,
  Array [
    11,
    Array [
      52,
      Array [
        Object {
          "date": "2020/12/23",
          "day": 23,
          "dayWeek": 3,
          "messagesAm": "msg1 23 dec 2020",
          "messagesPm": "off",
          "month": 11,
          "week": 52,
          "year": 2020,
        },
        Object {
          "date": "2020/12/24",
          "day": 24,
          "dayWeek": 4,
          "messagesAm": "off",
          "messagesPm": "msg1 24 dec 2020",
          "month": 11,
          "week": 52,
          "year": 2020,
        },
        Object {
          "date": "2020/12/25",
          "day": 25,
          "dayWeek": 5,
          "messagesAm": "off",
          "messagesPm": "off",
          "month": 11,
          "week": 52,
          "year": 2020,
        },
        Object {
          "date": "2020/12/26",
          "day": 26,
          "dayWeek": 6,
          "messagesAm": "off",
          "messagesPm": "off",
          "month": 11,
          "week": 52,
          "year": 2020,
        },
      ],
    ],
  ],
]
`;

exports[`report R5: groupByYear 2`] = `
Array [
  2020,
  Array [
    11,
    Array [
      53,
      Array [
        Object {
          "date": "2020/12/27",
          "day": 27,
          "dayWeek": 0,
          "messagesAm": "off",
          "messagesPm": "off",
          "month": 11,
          "week": 53,
          "year": 2020,
        },
        Object {
          "date": "2020/12/28",
          "day": 28,
          "dayWeek": 1,
          "messagesAm": "off",
          "messagesPm": "off",
          "month": 11,
          "week": 53,
          "year": 2020,
        },
        Object {
          "date": "2020/12/29",
          "day": 29,
          "dayWeek": 2,
          "messagesAm": "off",
          "messagesPm": "off",
          "month": 11,
          "week": 53,
          "year": 2020,
        },
        Object {
          "date": "2020/12/30",
          "day": 30,
          "dayWeek": 3,
          "messagesAm": "off",
          "messagesPm": "off",
          "month": 11,
          "week": 53,
          "year": 2020,
        },
        Object {
          "date": "2020/12/31",
          "day": 31,
          "dayWeek": 4,
          "messagesAm": "off",
          "messagesPm": "off",
          "month": 11,
          "week": 53,
          "year": 2020,
        },
      ],
    ],
  ],
]
`;

exports[`report R5: groupByYear 3`] = `
Array [
  2021,
  Array [
    0,
    Array [
      0,
      Array [
        Object {
          "date": "2021/01/01",
          "day": 1,
          "dayWeek": 5,
          "messagesAm": "off",
          "messagesPm": "off",
          "month": 0,
          "week": 0,
          "year": 2021,
        },
        Object {
          "date": "2021/01/02",
          "day": 2,
          "dayWeek": 6,
          "messagesAm": "off",
          "messagesPm": "off",
          "month": 0,
          "week": 0,
          "year": 2021,
        },
      ],
    ],
  ],
]
`;

exports[`report R5: groupByYear 4`] = `
Array [
  2021,
  Array [
    0,
    Array [
      1,
      Array [
        Object {
          "date": "2021/01/03",
          "day": 3,
          "dayWeek": 0,
          "messagesAm": "off",
          "messagesPm": "off",
          "month": 0,
          "week": 1,
          "year": 2021,
        },
        Object {
          "date": "2021/01/04",
          "day": 4,
          "dayWeek": 1,
          "messagesAm": "off",
          "messagesPm": "msg1 04 jan 2021",
          "month": 0,
          "week": 1,
          "year": 2021,
        },
      ],
    ],
  ],
]
`;

exports[`report R6.a: buildReport 1`] = `
"2020
[34mDec[39m, Week 51
[33mThu[39m, [33m17[39m
AM: [33mmsg1 17 dec 2020[39m
PM: [33moff[39m

[33mFri[39m, [33m18[39m
AM: [33mmsg1 18 dec 2020[39m
PM: [33moff[39m

[33mSat[39m, [33m19[39m
AM: [33mmsg1 19 dec 2020[39m
PM: [33moff[39m

[34mDec[39m, Week 52
[33mSun[39m, [33m20[39m
AM: [33mmsg1 20 dec 2020[39m
PM: [33moff[39m

[33mMon[39m, [33m21[39m
AM: [33mmsg1 21 dec 2020[39m
PM: [33moff[39m

[33mTue[39m, [33m22[39m
AM: [33mmsg1 22 dec 2020[39m
PM: [33moff[39m

[33mWed[39m, [33m23[39m
AM: [33moff[39m
PM: [33mmsg1 23 dec 2020[39m

[33mThu[39m, [33m24[39m
AM: [33moff[39m
PM: [33mmsg1 24 dec 2020[39m

[33mFri[39m, [33m25[39m
AM: [33moff[39m
PM: [33mmsg1 25 dec 2020[39m

[33mSat[39m, [33m26[39m
AM: [33moff[39m
PM: [33mmsg1 26 dec 2020[39m

[34mDec[39m, Week 53
[33mSun[39m, [33m27[39m
AM: [33moff[39m
PM: [33mmsg1 27 dec 2020[39m

[33mMon[39m, [33m28[39m
AM: [33moff[39m
PM: [33mmsg1 28 dec 2020[39m

[33mTue[39m, [33m29[39m
AM: [33moff[39m
PM: [33mmsg1 29 dec 2020[39m

[33mWed[39m, [33m30[39m
AM: [33moff[39m
PM: [33moff[39m

[33mThu[39m, [33m31[39m
AM: [33moff[39m
PM: [33moff[39m

2021
[34mJan[39m, Week 0
[33mFri[39m, [33m1[39m
AM: [33mmsg1 01 jan 2021[39m
PM: [33moff[39m

"
`;

exports[`report R6.b: buildReport last 2 days 1`] = `
"[31m2020[39m
[34mDec[39m
Week 53
[33mTue[39m, [33m29[39m
AM: [33moff[39m
PM: [33mmsg1 29 dec 2020[39m

[33mThu[39m, [33m31[39m
AM: [33moff[39m
PM: [33moff[39m

[31m2021[39m
[34mJan[39m
Week 0
[33mFri[39m, [33m1[39m
AM: [33mmsg1 01 jan 2021[39m
PM: [33moff[39m

"
`;

exports[`report R8: sendReport to console 1`] = `
"[33mWed[39m, [33m23[39m
AM: [33mmsg1 23 dec 2020[39m
PM: [33moff[39m

[33mThu[39m, [33m24[39m
AM: [33moff[39m
PM: [33mmsg1 24 dec 2020[39m

[33mFri[39m, [33m25[39m
AM: [33moff[39m
PM: [33moff[39m

[33mSat[39m, [33m26[39m
AM: [33moff[39m
PM: [33moff[39m

[34mDec[39m, Week 53
[33mSun[39m, [33m27[39m
AM: [33moff[39m
PM: [33moff[39m

[33mMon[39m, [33m28[39m
AM: [33moff[39m
PM: [33moff[39m

[33mTue[39m, [33m29[39m
AM: [33moff[39m
PM: [33moff[39m

[33mWed[39m, [33m30[39m
AM: [33moff[39m
PM: [33moff[39m

[33mThu[39m, [33m31[39m
AM: [33moff[39m
PM: [33moff[39m

2021
[34mJan[39m, Week 0
[33mFri[39m, [33m1[39m
AM: [33moff[39m
PM: [33moff[39m

[33mSat[39m, [33m2[39m
AM: [33moff[39m
PM: [33moff[39m

2021
[34mJan[39m, Week 1
[33mSun[39m, [33m3[39m
AM: [33moff[39m
PM: [33moff[39m

[33mMon[39m, [33m4[39m
AM: [33moff[39m
PM: [33mmsg1 04 jan 2021[39m

"
`;
